compact_str/src/features/rkyv.rs

```diff
@@ -0,0 +1,85 @@
+#![cfg_attr(docsrs, doc(cfg(feature = "rkyv")))]
+
+use rkyv::string::{
+    ArchivedString,
+    StringResolver,
+};
+use rkyv::{
+    Archive,
+    Deserialize,
+    DeserializeUnsized,
+    Fallible,
+    Serialize,
+    SerializeUnsized,
+};
+
+use crate::CompactString;
+
+impl Archive for CompactString {
+    type Archived = ArchivedString;
+    type Resolver = StringResolver;
+
+    #[inline]
+    unsafe fn resolve(&self, pos: usize, resolver: Self::Resolver, out: *mut Self::Archived) {
+        ArchivedString::resolve_from_str(self.as_str(), pos, resolver, out);
+    }
+}
+
+impl<S: Fallible + ?Sized> Serialize<S> for CompactString
+where
+    str: SerializeUnsized<S>,
+{
+    #[inline]
+    fn serialize(&self, serializer: &mut S) -> Result<Self::Resolver, S::Error> {
+        ArchivedString::serialize_from_str(self.as_str(), serializer)
+    }
+}
+
+impl<D: Fallible + ?Sized> Deserialize<CompactString, D> for ArchivedString
+where
+    str: DeserializeUnsized<str, D>,
+{
+    #[inline]
+    fn deserialize(&self, _: &mut D) -> Result<CompactString, D::Error> {
+        Ok(self.as_str().into())
+    }
+}
+
+impl PartialEq<CompactString> for ArchivedString {
+    #[inline]
+    fn eq(&self, other: &CompactString) -> bool {
+        PartialEq::eq(self.as_str(), other.as_str())
+    }
+}
+
+impl PartialOrd<CompactString> for ArchivedString {
+    #[inline]
+    fn partial_cmp(&self, other: &CompactString) -> Option<std::cmp::Ordering> {
+        PartialOrd::partial_cmp(self.as_str(), other.as_str())
+    }
+}
+
+#[cfg(test)]
+#[cfg_attr(miri, ignore)] // https://github.com/rust-lang/unsafe-code-guidelines/issues/134
```

Comment:
Does this stacked borrows violation come from our code, or rkyv's?

Reply:
rkyv. <https://github.com/rust-lang/unsafe-code-guidelines/issues/134#issuecomment-1087628968>

Actually, I think I see what the problem is in their code. In [repr.rs:40-62](https://github.com/rkyv/rkyv/blob/f552fabc204ebc27f812166e636a75d45d97cba1/rkyv/src/string/repr.rs#L40-L62) they would need to replace `self.inline.bytes.as_ptr()` with `self.inline.bytes[..].as_ptr()` to make MIRI understand what is going on. If I don't forget it, I'll try if that fixes the problem.